cleanrl/atari/v_c51_atari.py

```diff
@@ -0,0 +1,346 @@
+                env = gym.wrappers.RecordVideo(env, f"videos/{run_name}")
+        env = NoopResetEnv(env, noop_max=30)
+        env = MaxAndSkipEnv(env, skip=4)
+        env = EpisodicLifeEnv(env)
+        if "FIRE" in env.unwrapped.get_action_meanings():
+            env = FireResetEnv(env)
+        env = ClipRewardEnv(env)
+        env = gym.wrappers.ResizeObservation(env, (84, 84))
+        env = gym.wrappers.GrayScaleObservation(env)
+        env = gym.wrappers.FrameStack(env, 4)
+        env.seed(seed)
+        env.action_space.seed(seed)
+        env.observation_space.seed(seed)
+        return env
+
+    return thunk
+
+envs = gym.vector.SyncVectorEnv(
+        [make_env(args.gym_id, args.seed + i, i, args.capture_video, run_name) for i in range(args.num_envs)]
+    )
+assert isinstance(envs.single_action_space, gym.spaces.Discrete), "only discrete action space is supported"
+
+# TRY NOT TO MODIFY: seeding
+device = torch.device('cuda' if torch.cuda.is_available() and args.cuda else 'cpu')
+random.seed(args.seed)
+np.random.seed(args.seed)
+torch.manual_seed(args.seed)
+torch.backends.cudnn.deterministic = args.torch_deterministic
+
+
+def layer_init(layer, std=np.sqrt(2), bias_const=0.0):
+    torch.nn.init.orthogonal_(layer.weight, std)
+    torch.nn.init.constant_(layer.bias, bias_const)
+    return layer
+
+class QNetwork(nn.Module):
+    def __init__(self, envs, device, frames=4, n_atoms=51, v_min=-10, v_max=10):
+        super(QNetwork, self).__init__()
+        self.n_atoms = n_atoms
+        self.atoms = torch.linspace(v_min, v_max, steps=n_atoms).to(device)
+        self.network = nn.Sequential(
+            nn.Conv2d(frames, 32, 8, stride=4),
+            nn.ReLU(),
+            nn.Conv2d(32, 64, 4, stride=2),
+            nn.ReLU(),
+            nn.Conv2d(64, 64, 3, stride=1),
+            nn.ReLU(),
+            nn.Flatten(),
+            nn.Linear(3136, 512),
+            nn.ReLU(),
+            nn.Linear(512, envs.single_action_space.n * n_atoms)
+        )
+
+    def forward(self, x, device):
+        #x = torch.Tensor(x).to(device)
+        return self.network(x/255.0)
+
+    def get_action(self, envs, device, x,action=None):
+        logits = self.forward(x, device)
+        # probability mass function for each action
+        pmfs = torch.softmax(logits.view(len(x), envs.single_action_space.n, self.n_atoms), dim=2)
+        q_values = (pmfs*self.atoms).sum(2)
+        if action is None:
+            action = torch.argmax(q_values, 1)
+            random_action = torch.randint(0, envs.single_action_space.n, (len(x),), device=device)
+            random_action_flag = torch.rand(len(x), device=device) > epsilon
+            action = torch.where(random_action_flag, action, random_action)
+           
+        return action, pmfs[torch.arange(len(x)), action]
+
+def linear_schedule(start_e: float, end_e: float, duration: int, t: int):
+    slope =  (end_e - start_e) / duration
+    return max(slope * t + start_e, end_e)
```

Comment:
Please move the Agent class and functions at the top level of the script

Reply:
Updated